### DISCUSSION TITLE
Error: index row size exceeds btree version 4 maximum for index

### DISCUSSION THREAD

**Discussion by pashkinelfe** [supabase/supabase#14898] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/error-index-row-size-exceeds-btree-version-4-maximum-for-index-LMmoeU)._

## Error

    index row size exceeds btree version 4 maximum 2704 for index "idx_name"

## Summary

PG has a limit on a B-tree tuple(=row) size. It needs to fit at least 3 B-tree tuples on a 8Kb page. That could not be changed.

B-tree row can be a single attribute or multiple attributes. These cases are better addressed separately.

## B-tree is built with multiple attributes

B-tree with multiple attributes will perform better than several only in case the likely SELECT queries use several attributes that include the first attributes that are in the index. I.e. select by 1-st, 2-nd, 3-d but not by 2-nd, 3-d and 5-th index attributes.

The other case when multiple attributes B-tree is good is when we have INSERT/UPDATE workload that is comparable to SELECT load (generally SELECTS a way more often). Then we can save speed-up updating a single index instead of several at INSERT/UPDATE at cost of SELECT performance decrease.

But most likely we have multiple attributes B-tree index due to some automation tool, not by intention. Even without the mentioned error it's best to build separate single-attribute indexes for each attribute from it. Then drop multiple attributes B-tree index. This is a must and the only solution when we have this error though.

## B-tree is built on a single attribute that is very long

This can be if the index is built on text, JSON column etc. It's not prohibited to build B-tree on these datatypes, but it's also ineffective. Why?

One of the measures of index efficiency is the ratio of index entries to the width of all possible values space for this datatype. If we have say 100000 distinct values of int32 in the index then the ratio is 1/40000. If we have text with length of 2704 bytes (maximum for B-tree index) we can hardly imagine the number of distinct values that gives us even a comparable ratio. That said indexing of that long values stores much redundancy in the index.

The solution: use some kind of hashing to transfer the values to much narrower space. I.e. md5. Build a functional index (=index by expression):

```sql
create index on table_name (MD5(column_name));
```

instead of:

```sql
create index on table_name (column_name);
```

Then you must modify you SELECTs to be using the same function (otherwise the functional index will not be used in SELECT queries). I.e.

```sql
select * from table_name where MD5(column_name) = MD5('search_value');
```

instead of

```sql
select * from table_name where column_name = 'search_value';
```

[More on building index by expression](https://www.postgresql.org/docs/current/sql-createindex.html)

For some non-text datatypes that support queries by partial inclusion—such as checking whether a key-value pair exists in JSON, or implementing `tsvector` phrase search—use GiST or GIN indexes. These index types operate on a narrower value space than the full content being indexed.

[More on GIN/GiST indexes](https://www.postgresql.org/docs/15/textsearch-indexes.html)

**Comment by cmxpro** [comment]
This is helpful! 

I encountered the same `index row size` issue with a UNIQUE index. Thus, I instead created a UNIQUE index on the md5-hashed value of the original column.

`CREATE UNIQUE INDEX my_table_my_column_md5_key ON public.my_table USING btree (md5(my_column))`

My challenge was that I also wanted to use the `ON CONFLICT DO NOTHING` statement with this index when inserting (upserting) data into this table.  

Here are the challenges I encountered:

- SQL supports an index_expression as the ON CONFLICT `conflict_target` (see [documentation](https://www.postgresql.org/docs/current/sql-insert.html)), as long as the index_expression is immutable (which is the case for md5()). Example: 
`INSERT INTO my_table (my_column)
VALUES ('some value'), ('some value')
ON CONFLICT ((md5(my_column))) DO NOTHING;`

- However, neither supabase-js nor PostgREST directly seem to support index_expressions with the ON CONFLICT statement. 

- Thus, creating a stored procedure and calling this e.g. via `const { data, error } = await supabase.rpc('my_stored_procedure')` seems to be the best solution in this case. It would be great to see PostgREST (and supabase-js) extended to support index_expressions with the ON CONFLICT statement in the future.

**Reply by steve-chavez** [reply]
Noted the use case on https://github.com/PostgREST/postgrest/issues/2123#issuecomment-1604405626
